lite/verifying/provider.go

```diff
@@ -0,0 +1,440 @@
+		lite.NewDBProvider(lvlDBFile, dbm.NewDB(dbName, dbm.LevelDBBackend, rootDir)),
+	)
+
+	return makeProvider(chainID, options.TrustPeriod, trust, lclient.NewProvider(chainID, client), logger)
+}
+
+// NOTE If you retain the resulting verifier in memory for a long time,
+// usage of the verifier may eventually error, but immediate usage should
+// not error like that, so that e.g. cli usage never errors unexpectedly.
+func NewProvider(chainID, rootDir string, client lclient.SignStatusClient, logger log.Logger, cacheSize int, options TrustOptions) (*Provider, error) {
+
+	vp := initProvider(chainID, rootDir, client, logger, cacheSize, options)
+
+	// Get the latest source commit, or the one provided in options.
+	trustCommit, err := getTrustCommit(client, options)
+	if err != nil {
+		return nil, err
+	}
+
+	err = vp.fillValsetAndSaveFC(trustCommit, nil, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	// sanity check
+	if time.Now().Sub(trustCommit.Time) <= 0 {
+		panic(fmt.Sprintf("impossible time %v vs %v", time.Now(), trustCommit.Time))
+	}
+
+	// Otherwise we're syncing within the unbonding period.
+	// NOTE: There is a duplication of fetching this latest commit (since
+	// UpdateToHeight() will fetch it again, and latestCommit isn't used), but
+	// it's only once upon initialization of a validator so it's not a big
+	// deal.
+	if options.TrustHeight > 0 {
+		latestCommit, err := client.Commit(nil)
+		if err != nil {
+			return nil, err
+		}
+		err = vp.UpdateToHeight(chainID, latestCommit.SignedHeader.Height)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return vp, nil
+}
+
+// getTrustCommit returns a commit trusted with weak subjectivity. It either:
+// 1. Fetches a commit at height provided in options and ensures the specified commit
+// 	is within the trust period of latest block
+// 2. Trusts the remote node and gets the latest commit
+// 3. Returns an error if the height provided in trust option is too old to sync to latest.
+func getTrustCommit(client lclient.SignStatusClient, options TrustOptions) (types.SignedHeader, error) {
+
+	// Get the lastest commit always
+	latestBlock, err := client.Commit(nil)
+	if err != nil {
+		return types.SignedHeader{}, err
+	}
+
+	// If the user has set a root of trust, confirm it then update to newest
+	if options.TrustHeight != 0 && options.TrustHash != nil {
+		trustBlock, err := client.Commit(&options.TrustHeight)
+		if err != nil {
+			return types.SignedHeader{}, err
+		}
+
+		if latestBlock.Time.Sub(trustBlock.Time) > options.TrustPeriod {
+			return types.SignedHeader{}, fmt.Errorf("Your Trusted Block Height is older than the trust period from Latest Block")
+		}
+
+		trustCommit := trustBlock.SignedHeader
+		if !bytes.Equal(trustCommit.Hash(), options.TrustHash) {
+			return types.SignedHeader{}, fmt.Errorf("WARNING!!! Expected height/hash %v/%X but got %X",
+				options.TrustHeight, options.TrustHash, trustCommit.Hash())
```

Comment:
The error messages reads more like a log output. Remove the "WARNING!!!" at least?